### PR TITLE
Refactor duplicate __sleep and __wakeup methods

### PR DESCRIPTION
**Pull request by aashan10**
partial fix for #8917

### CODE REVIEW

File: models/Element/AbstractElement.php [hunk 1]
```diff
@@ -479,12 +487,51 @@ public function getVersions()
...
+
         $parentVars = parent::__sleep();
-        $blockedVars = ['dependencies'];
+
+        $type = Service::getElementType($this);
+
+        // this should be removed when the o_ prefix will be removed from data objects
+        $hasPrefix = in_array($type, ['document', 'asset']);
+
+        $blockedVars = match ($type) {
```
Comment (fashxp) [inline models/Element/AbstractElement.php]: I would add a property to the `AbstractElement` class which is overwritten for every subclass then ... 


In `AbstractElement`: 
```php
protected $blockedVars = [];
```

In `Document`: 
```php
protected $blockedVars = ['hasChildren', 'versions', 'scheduledTasks', 'parent', 'fullPathCache'];
```

File: models/Document.php [hunk 2]
```diff
@@ -1272,22 +1272,6 @@ public function setParent($parent)
...
-            $this->removeInheritedProperties();
-        } else {
-            // this is if we want to cache the object
-            $blockedVars = array_merge($blockedVars, ['children', 'properties']);
-        }
-
-        return array_diff($parentVars, $blockedVars);
-    }
-
     public function __wakeup()
```
Comment (fashxp) [inline models/Document.php]: is there a reason why this method is still here?

File: models/Element/AbstractElement.php (hunk 1 above)
Comment (fashxp) [reply models/Element/AbstractElement.php]: or maybe we even should introduce a protected method `getBlockedVars` that also returns `(o_)children` etc.

**Review by fashxp — CHANGES_REQUESTED** [review]

File: models/Element/AbstractElement.php (hunk 1 above)
Comment (aashan10) [reply models/Element/AbstractElement.php]: I have added a protected variable `$blockedVars` and then added a method `getBlockedVars` for all the logic

File: models/Document.php (hunk 2 above)
Comment (aashan10) [reply models/Document.php]: no, had missed this one for some reason. I've fixed and updated

File: models/Asset.php [hunk 3]
```diff
@@ -215,6 +215,11 @@ class Asset extends Element\AbstractElement
      */
     protected $versionCount = 0;
 
+    /**
+     * @var array
+     */
+    protected array $blockedVars = ['scheduledTasks', 'hasChildren', 'versions', 'parent', 'stream', 'children', 'properties'];
+
```
Comment (fashxp) [inline models/Asset.php]:
```suggestion
    /**
     * @return array
     */
    protected function getBlockedVars(): array
    {

        $blockedVars = ['scheduledTasks', 'hasChildren', 'versions', 'parent', 'stream'];

        if (! $this->isInDumpState()) {
            // this is if we want to cache the asset
            $blockedVars = array_merge($blockedVars, ['children', 'properties']);
        }
        
        return $blockedVars; 

    }
```

File: models/DataObject/AbstractObject.php [hunk 4]
```diff
@@ -225,6 +218,11 @@ abstract class AbstractObject extends Model\Element\AbstractElement
      */
     protected $o_versionCount = 0;
 
+    /**
+     * @var array
+     */
+    protected array $blockedVars = ['o_hasChildren', 'o_parent', 'scheduledTasks', 'o_versions', 'omitMandatoryCheck', 'dependencies'];
+
```
Comment (fashxp) [inline models/DataObject/AbstractObject.php]:
```suggestion
    /**
     * @return array
     */
    protected function getBlockedVars(): array
    {
        $blockedVars = ['o_hasChildren', 'o_versions', 'o_class', 'scheduledTasks', 'o_parent', 'omitMandatoryCheck'];

        if ($this->isInDumpState()) {
            // this is if we want to make a full dump of the object (eg. for a new version), including children for recyclebin
            $blockedVars = array_merge($blockedVars, ['o_dirtyFields']);
        } else {
            // this is if we want to cache the object
            $blockedVars = array_merge($blockedVars, ['o_children', 'properties']);
        }
        
        return $blockedVars; 
    }

```

File: models/Document.php [hunk 5]
```diff
@@ -189,6 +189,11 @@ class Document extends Element\AbstractElement
      */
     protected $versionCount = 0;
 
+    /**
+     * @var array
+     */
+    protected array $blockedVars = ['hasChildren', 'parent', 'scheduledTasks', 'versions', 'fullPathCache'];
+
```
Comment (fashxp) [inline models/Document.php]:
```suggestion
    /**
     * @return array
     */
    protected function getBlockedVars(): array
    {
        $blockedVars = ['hasChildren', 'versions', 'scheduledTasks', 'parent', 'fullPathCache'];

        if (! $this->isInDumpState()) {
            // this is if we want to cache the object
            $blockedVars = array_merge($blockedVars, ['children', 'properties']);
        }
        
        return $blockedVars; 
    }

```

File: models/Element/AbstractElement.php [hunk 6]
```diff
@@ -55,6 +56,19 @@ abstract class AbstractElement extends Model\AbstractModel implements ElementInt
      */
     protected ?array $properties = null;
 
+    /**
+     * @internal
+     * @var array
+     */
+    protected array $blockedVars = [];
```
Comment (fashxp) [inline models/Element/AbstractElement.php]:
```suggestion

```

File: models/Element/AbstractElement.php [hunk 7]
```diff
@@ -474,15 +488,49 @@ public function getVersions()
...
+            $blockedVars = [...$this->blockedVars, $hasPrefix ? 'o_dirtyFields' : 'dirtyFields'];
+            $this->removeInheritedProperties();
+
+            return $blockedVars;
+        }
+
+        return [...$this->blockedVars, $hasPrefix ? 'o_children' : 'children', 'properties'];
+
+    }
+
```
Comment (fashxp) [inline models/Element/AbstractElement.php]:
```suggestion
```

File: models/Element/AbstractElement.php [hunk 8]
```diff
@@ -474,15 +488,49 @@ public function getVersions()
...
     /**
      * {@inheritdoc}
      */
     public function __sleep()
     {
-        $parentVars = parent::__sleep();
-        $blockedVars = ['dependencies'];
+        return array_diff(parent::__sleep(), $this->getBlockedVars());
```
Comment (fashxp) [inline models/Element/AbstractElement.php:687]:
```suggestion

        if ($this->isInDumpState()) {
            // this is if we want to make a full dump of the object (eg. for a new version), including children for recyclebin
            $this->removeInheritedProperties();
        }

        return array_diff(parent::__sleep(), $this->getBlockedVars());
```

File: models/Asset.php [hunk 9]
```diff
@@ -1842,40 +1847,6 @@ public function setParent($parent)
...
-            $this->removeInheritedProperties();
-        } else {
-            // this is if we want to cache the asset
-            $blockedVars = array_merge($blockedVars, ['children', 'properties']);
-        }
-
-        return array_diff($parentVars, $blockedVars);
-    }
-
-    public function __wakeup()
```
Comment (fashxp) [inline models/Asset.php:1718]: we cannot remove that one, because this is different to the one in the `AbstractElement`

**Review by fashxp — CHANGES_REQUESTED** [review]
Hmm, the introduced changes were not exactly what I meant and also would change behavior ... please have a look at my suggestions below... 

please also resolve the conflicts. 

thx

File: models/Element/AbstractElement.php (hunk 8 above)
Comment (aashan10) [reply models/Element/AbstractElement.php:687]: I have made changes according to the review

File: models/Asset.php (hunk 3 above)
Comment (aashan10) [reply models/Asset.php]: This has been changed

File: models/Element/AbstractElement.php [hunk 10]
```diff
@@ -662,15 +662,40 @@ public function getVersions()
         return [];
     }
 
+    protected function getBlockedVars(): array
+    {
+        return ['dependencies'];
```
Comment (fashxp) [inline models/Element/AbstractElement.php]: 'parent' got lost now again here?

**Review by fashxp — CHANGES_REQUESTED** [review]

**Review by fashxp — APPROVED** [review]

**Pull request merged**